ciao/arch/architecture.md

```diff
@@ -0,0 +1,422 @@
+The image depicts a cluster with one master node, one network node and two
+compute nodes.  The master node runs the scheduler and controller services.
+Launcher runs on both the compute nodes and the network node.  There are four
+compute instances, two on each compute node, and one CNCI.  Inside the
+CNCI the final service, the cnci-agent, runs.  All the services communicate with
+each other via the scheduler using the SSNTP protocol.  Controller, the VM
+compute instance and the CNCI also communicate with the ceph cluster.  The
+rootfs images for the VM instance and the CNCI were originally created by
+controller and are stored in ceph.  One of the docker instances also needs to
+communicate with the ceph cluster as it has a data volume attached to it.
+Finally, the user inspects and manipulates the cluster using the ciao tool which
+executes the user's commands by communicating over HTTPS via the controller.
+
+We'll take a closer look at the CIAO services, their communication protocol and the
+user focused ciao tool in the next few sections.
+
+
+### The ciao tool
+
+Users and administrators interact with the cluster using a command line tool
+called ciao.  This tool can be used to perform all the available operations on
+the cluster.  The ciao tool itself is little more than a front end that maps
+users' requests onto a REST API and sends these requests to controller.  The
+ciao tool and controller communicate via HTTPS only.  For examples of some ciao
+commands please see [The command line](ciao.html).
+
+### The controller
+
+The controller is the central management entity for the cluster.  It provides a
+REST API that can be used to inspect and manipulate the cluster, it processes
+and validates users' requests received via this API and it maintains the cluster
+state.  Controller keeps track of the cluster's tenants, workloads, images,
+instances and volumes.  Although the controller can process some commands it
+receives by itself, such as creating a new image, it needs help fulfilling other
+commands such as starting or stopping an instance or creating a tenant.  To
+perform these other actions controller needs to communicate with the other
+services in the cluster's control plane.
+
+### Launcher
+
+One Launcher instance runs on each compute and network node.  It is responsible
+for creating and managing the lifecycles of compute and CNCI instances.  It also
+monitors the state of the node on which it runs and the instances it manages
+sending information about these entities back to controller and scheduler on a
+regular basis.  Launcher keeps track of the available resources on its node, and
+if it determines that resources are too low, will report to scheduler that it
+is full and cannot host any new instances for the time being.
+
+### The scheduler
+
+The scheduler serves three purpose.
+
+1. It acts as the communication gateway for the cluster's control plane.  It routes
```

Comment:
Again I wouldn't use gateway as in networking it's a communication between two different networks which is not the case. If anything is a gateway it is controller..or the CNCI agent.